File: mobile/ios/Sources/Store/OutboundQueue.swift
import Foundation

public struct PendingOp: Codable, Sendable, Equatable, Identifiable {
    public enum Kind: Codable, Sendable, Equatable {
        case completeTask(String)
    }

    public let id: UUID
    public let kind: Kind
    public let createdAt: Date
    public var attempts: Int

    public init(id: UUID = UUID(), kind: Kind, createdAt: Date = Date(), attempts: Int = 0) {
        self.id = id
        self.kind = kind
        self.createdAt = createdAt
        self.attempts = attempts
    }
}

public struct OutboundQueue: Sendable {
    public static let maxAttempts = 5
    private let file: URL

    public init(directory: URL = SharedContainer.url) {
        self.file = directory.appending(path: "outbox.json")
    }

    public func all() -> [PendingOp] {
        guard let data = try? Data(contentsOf: file) else { return [] }
        return (try? ProtocolDecoder.json.decode([PendingOp].self, from: data)) ?? []
    }

    public func enqueue(_ op: PendingOp) throws {
        var ops = all()
        // Tapping the same row twice is one intent, not two.
        guard !ops.contains(where: { $0.kind == op.kind }) else { return }
        ops.append(op)
        try write(ops)
    }

    public func remove(id: UUID) throws {
        try write(all().filter { $0.id != id })
    }

    public func recordAttempt(id: UUID) throws {
        var ops = all()
        guard let index = ops.firstIndex(where: { $0.id == id }) else { return }
        ops[index].attempts += 1
        // An op that has failed this many times is not going to start working.
        // Dropping it is better than a queue that retries forever on every
        // background wake.
        if ops[index].attempts >= Self.maxAttempts { ops.remove(at: index) }
        try write(ops)
    }

    private func write(_ ops: [PendingOp]) throws {
        try ProtocolDecoder.jsonEncoder.encode(ops).write(to: file, options: .atomic)
    }
}

File: mobile/android/app/src/main/kotlin/sh/openagi/mobile/store/OutboundQueue.kt
package sh.openagi.mobile.store

import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import sh.openagi.mobile.protocol.InstantSerializer
import sh.openagi.mobile.protocol.ProtocolJson
import java.io.File
import java.time.Instant
import java.util.UUID

@Serializable
data class PendingOp(
    val id: String,
    val kind: Kind,
    @Serializable(with = InstantSerializer::class) val createdAt: Instant,
    val attempts: Int = 0,
) {
    @Serializable
    sealed class Kind {
        @Serializable
        data class CompleteTask(val taskId: String) : Kind()
    }

    companion object {
        fun completeTask(taskId: String): PendingOp =
            PendingOp(UUID.randomUUID().toString(), Kind.CompleteTask(taskId), Instant.now())
    }
}

class OutboundQueue(directory: File) {
    private val file = File(directory, "outbox.json")
    private val lock = Any()

    fun all(): List<PendingOp> = synchronized(lock) { allLocked() }

    private fun allLocked(): List<PendingOp> {
        if (!file.exists()) return emptyList()
        return try {
            ProtocolJson.json.decodeFromString(ListSerializer(PendingOp.serializer()), file.readText())
        } catch (error: Exception) {
            emptyList()
        }
    }

    fun enqueue(op: PendingOp) {
        synchronized(lock) {
            val ops = allLocked()
            // Tapping the same row twice is one intent, not two.
            if (ops.any { it.kind == op.kind }) return
            writeLocked(ops + op)
        }
    }

    fun remove(id: String) {
        synchronized(lock) { writeLocked(allLocked().filterNot { it.id == id }) }
    }

    fun recordAttempt(id: String) {
        synchronized(lock) {
            val ops = allLocked().toMutableList()
            val index = ops.indexOfFirst { it.id == id }
            if (index < 0) return
            val bumped = ops[index].copy(attempts = ops[index].attempts + 1)
            // An op that has failed this many times is not going to start working.
            // Dropping it is better than a queue that retries forever on every
            // background wake.
            if (bumped.attempts >= MAX_ATTEMPTS) ops.removeAt(index) else ops[index] = bumped
            writeLocked(ops)
        }
    }

    private fun writeLocked(ops: List<PendingOp>) {
        val encoded = ProtocolJson.json.encodeToString(ListSerializer(PendingOp.serializer()), ops)
        val temp = File(file.parentFile, "outbox.json.tmp")
        temp.writeText(encoded)
        temp.renameTo(file)
    }

    companion object {
        const val MAX_ATTEMPTS = 5
    }
}
